redwood-layout-uiview/src/commonMain/kotlin/app/cash/redwood/layout/uiview/YogaUIView.kt

```diff
@@ -0,0 +1,180 @@
+
+import app.cash.redwood.layout.api.Constraint
+import app.cash.redwood.yoga.MeasureCallback
+import app.cash.redwood.yoga.MeasureMode
+import app.cash.redwood.yoga.Node
+import app.cash.redwood.yoga.Size
+import kotlinx.cinterop.CValue
+import kotlinx.cinterop.cValue
+import kotlinx.cinterop.useContents
+import platform.CoreGraphics.CGRectMake
+import platform.CoreGraphics.CGRectZero
+import platform.CoreGraphics.CGSize
+import platform.CoreGraphics.CGSizeMake
+import platform.UIKit.UIView
+import platform.UIKit.UIViewNoIntrinsicMetric
+
+internal class YogaUIView : UIView(cValue { CGRectZero }) {
+  val rootNode = Node()
+
+  var width = Constraint.Wrap
+  var height = Constraint.Wrap
+
+  var applyModifier: (Node, Int) -> Unit = { _, _ -> }
+
+  override fun intrinsicContentSize(): CValue<CGSize> {
+    return calculateLayoutWithSize(Size(Size.Undefined, Size.Undefined)).toCGSize()
+  }
+
+  override fun sizeThatFits(size: CValue<CGSize>): CValue<CGSize> {
+    val bounds = size.useContents { sizeToBounds(this) }
+    val output = calculateLayoutWithSize(bounds)
+    return output.toCGSize()
+  }
+
+  override fun layoutSubviews() {
+    val bounds = bounds.useContents { sizeToBounds(size) }
+    calculateLayoutWithSize(bounds)
+
+    for (childNode in rootNode.children) {
+      YGApplyLayoutToViewHierarchy(childNode)
+    }
+  }
+
+  private fun calculateLayoutWithSize(size: Size): Size {
+    YGAttachNodesFromViewHierachy(this)
+
+    for ((index, node) in rootNode.children.withIndex()) {
+      applyModifier(node, index)
+    }
+
+    rootNode.measure(size.width, size.height)
+
+    return Size(rootNode.width, rootNode.height)
+  }
+
+  private fun sizeToBounds(size: CGSize): Size {
+    return Size(
+      width = sizeToBoundsDimension(width, size.width),
+      height = sizeToBoundsDimension(height, size.height),
+    )
+  }
+
+  private fun sizeToBoundsDimension(constraint: Constraint, dimension: Double): Float {
+    if (constraint == Constraint.Wrap || dimension == UIViewNoIntrinsicMetric) {
+      return Size.Undefined
+    } else {
+      return dimension.toFloat()
+    }
+  }
+}
+
+private fun YGAttachNodesFromViewHierachy(yoga: YogaUIView) {
+  if (yoga.typedSubviews.isEmpty()) {
+    yoga.rootNode.children.clear()
+    yoga.rootNode.measureCallback = ViewMeasureCallback(yoga)
+    return
+  }
+
+  // Nodes with children cannot have measure functions.
+  yoga.rootNode.measureCallback = null
+
+  val currentViews = yoga.rootNode.children.mapNotNull { it.view }
+  val subviews = yoga.typedSubviews
+  if (currentViews != subviews) {
+    yoga.rootNode.children.clear()
+    for (view in subviews) {
+      yoga.rootNode.children += view.asNode()
+    }
+  }
+}
+
+private class ViewMeasureCallback(val view: UIView) : MeasureCallback {
+  override fun measure(
+    node: Node,
+    width: Float,
+    widthMode: MeasureMode,
+    height: Float,
+    heightMode: MeasureMode,
+  ): Size {
+    val view = node.view!!
```

Comment:
This hides the class property of the same name. I think we only need either this or the class property but not both.

Reply:
Ah yep - good catch